src/math/nfft/dnfft_1d_kernels.cu

```diff
@@ -0,0 +1,168 @@
+// Copyright (C) 2018 ETH Zurich
+// Copyright (C) 2018 UT-Battelle, LLC
+// All rights reserved.
+//
+// See LICENSE.txt for terms of usage.
+// See CITATION.txt for citation guidelines if you use this code for scientific publications.
+//
+// Author: Giovanni Balduzzi (gbalduzz@itp.phys.ethz.ch)
+//
+// This file implements the GPU kernels used by the Dnfft1D class.
+
+#include "dca/math/nfft/kernels_interface.hpp"
+
+#include <array>
+#include <cuda_runtime.h>
+
+#include "dca/math/nfft/nfft_helper.cu.hpp"
+#include "dca/linalg/util/atomic_add_cuda.cu.hpp"
+#include "dca/util/integer_division.hpp"
+
+namespace dca {
+namespace math {
+namespace nfft {
+namespace details {
+// dca::math::nfft::details::
+
+std::array<int, 2> getBlockSize(const int ni, const int block_size) {
+  const int n_threads = std::min(block_size, ni);
+  const int n_blocks = util::ceilDiv(ni, n_threads);
+  return std::array<int, 2>{n_blocks, n_threads};
+}
+
+std::array<dim3, 2> getBlockSize(const uint i, const uint j, const uint block_size = 32) {
+  const uint n_threads_i = std::min(block_size, i);
+  const uint n_threads_j = std::min(block_size, j);
+  if (n_threads_i * n_threads_j > 32 * 32)
+    throw(std::logic_error("Block size is too big"));
+
+  const uint n_blocks_i = dca::util::ceilDiv(i, n_threads_i);
+  const uint n_blocks_j = dca::util::ceilDiv(j, n_threads_j);
+
+  return std::array<dim3, 2>{dim3(n_blocks_i, n_blocks_j), dim3(n_threads_i, n_threads_j)};
+}
+
+template <typename ScalarType>
+struct HelperSelector {
+  static NfftHelperManager<ScalarType> value;
+};
+template <typename ScalarType>
+NfftHelperManager<ScalarType> HelperSelector<ScalarType>::value;
+
+struct ConfigElem {
```

Comment:
I am not sure why I need to redefine `ConfigElem` even though it should be defined and in scope from `kernels_interface.hpp`, but this is the only way it compiles. Any idea on the cause?

Reply:
Solved: DCA_HAVE_CUDA was not defined.

Reply:
The *.cu files are compiled with different options/flags if CUDA_PROPAGATE_HOST_FLAGS is not on. the easiest fix is to just #include dca/config/haves_defines so that the right things are picked up.